### ISSUE TITLE
Lots of tests fail when running with workers > 1 - Using default chrome headless runner

### ISSUE THREAD

**Issue by massimocode** [wallabyjs/public#3397]
#### Summary of issue
Hey there, long-time Wallaby user here, since the early days in mid 2015 and massive fan.

I have a frontend project which I'm testing using Wallaby's default headless chrome runner. There is general flakiness, in that some tests will randomly fail across different test suites. I noticed that when I set workers initial/regular to 1, then the flakiness disappears. If I set it to anything higher than that (i.e. 2 or more) or I remove the workers section of the configuration, then the flakiness is there.

I do not understand why there would be any flakiness as I am not aware of any global state that could cause this. I don't believe the flakiness is due to lack of system resources, as I recently upgraded to a faster machine and the flakiness only got much worse when I did that. It should also be noted that on my previous machine I was using Windows 10 and had configured Wallaby to use a standalone fixed version of Chrome (108). This was much less flaky than my current machine which is using Windows 11 and using my machine's Chrome installation (the latest Chrome, version 128).

So I'd like to know if any global state is shared between these parallel workers. My other assumption is that perhaps Chrome has some performance optimisations where "background" instances don't work like "foreground" instances, so when using more than 1 worker then some workers are running "differently" (i.e. throttled or in some kind of memory / energy saver mode).

Any sort of help to pinpoint and resolve this issue would be greatly appreciated because I rely on Wallaby as my only test runner.

### Wallaby diagnostics report (set to use 1 worker only)
[diagnostics-1-worker-restriction.txt](https://github.com/user-attachments/files/17003971/diagnostics-1-worker-restriction.txt)

### Wallaby diagnostics report (without the 1 worker restriction)
[diagnostics-no-worker-restriction.txt](https://github.com/user-attachments/files/17003972/diagnostics-no-worker-restriction.txt)

**Comment by massimocode** [comment]
As an update - I tried electron with workers limited to 1 and this was very flaky, approximately equivalent to chrome when not limited to 1.

**Comment by massimocode** [comment]
I just downloaded Chrome 108 and configured Wallaby to use that instead. The flakiness is much better. Only around 1% of tests fail when running all project tests, as opposed to approximately 25% of tests.

**Comment by massimocode** [comment]
I downloaded Chrome 121 and that was fine (only minimally flaky). But 128 has the very flaky issue. I'm going to try and see what version of Chrome has introduced this issue (things becoming stupidly flaky).

**Comment by smcenlly** [comment]
Thanks for reporting your issue, and for the diagnostic reports.

You mentioned that tests sometimes randomly fail. When they fail, what is the reason for the failure (timeout, invalid assertion, etc.)?

When Wallaby runs your tests with Chrome using multiple workers, it launches a single chrome instance and then opens a tab for each test run. These *should* be isolated unless (as you suggested) Chrome is doing something different when multiple tabs are being opened.

We haven't seen any flakiness using the latest version of Chrome, but we will test on some larger projects today to try and reproduce your issue. I'm wondering if there's something about your test / project structure that is causing the flakiness. **Are you able to share your repo, or provide a sample repo that has the same problem?** If so, you can either email us or share a private repo.

**Comment by massimocode** [comment]
Thanks for the response.

These are essentially UI tests. They basically render the application on the DOM, then they interact with the application via the DOM and then assertvia querying the DOM that the application has reacted to those interactions as expected.

These UI tests are similar to the type of tests typically run using Selenium / Webdriver, so it's expected that there could be some flakiness due to timing issues, but it should be very rare and minimal.

However, there is also a test suite which often fails which is not a UI test and does not involve the DOM, it's simply a series of tests for an async queue. So I suspect something fishy is going on with regard to how tests that aren't in the foreground tab run. Is there a wallaby configuration that I can set which will force wallaby to use separate chrome windows for each worker?

Regardless, I'm happy to send you the source code for you to replicate it. It's a closed-source application. Please provide an email address where I can send it to.

**Comment by smcenlly** [comment]
> Is there a wallaby configuration that I can set which will force wallaby to use separate chrome windows for each worker?

Not at the moment. From what I understand of Chrome design, each tab is running in its own process anyway (unless that's changed). If we can reproduce your problem, we can see if that makes a difference.

Our email address is hello@wallabyjs.com

**Comment by massimocode** [comment]
I've emailed it over with reference ec983f

The encryption key for the 7zip file is 9c5eab8ed3974004989eb83c02bea95d

You'll need to extract it (using the key above) then you'll need to npm install in the following directories:
- api (possibly optional)
- api-shared (possibly optional)
- frontend (required)

Then run wallaby JS, picking the `frontend/wallaby.conf.js` file

By default, it will use the latest chrome installed. On my machine that's v128 which is incredibly flaky. When I manually specify to use an earlier standalone chromium version then things are as I would expect them to be.

There ae 405 tests in this example repo. They should all pass. It's possible that a small handful (~1%) might fail, rerunning them makes them pass. What shouldn't happen is for a much larger ratio of them to fail.

Looking forward to hearing if you are able to replicate this.

**Comment by smcenlly** [comment]
We found and fixed a Chrome related bug in Wallaby that may have been affecting you (we're not sure). Could you please also try updating to Wallaby core `v1.0.1599` and let us know if that fixes the Chrome ` 128.0.6613.138` behavior for you?

Unfortunately the password isn't working for the archive. We've tried on a couple of machines to decrypt it, but it says the password is incorrect. If you're still having the problem, can you please provide us with a new archive after confirming it's extracting for you?

**Comment by smcenlly** [comment]
> Unfortunately the password isn't working for the archive. We've tried on a couple of machines to decrypt it, but it says the password is incorrect. If you're still having the problem, can you please provide us with a new archive after confirming it's extracting for you?

Please disregard, we were able to get it working.

**Comment by smcenlly** [comment]
When running your tests, your tests were also flakey when we ran them. 

When using the latest version of Chrome, we also had a problem. We could see this was related to your `test/test-environment/test-environment.ts` `waitUntil` function. If we increased the timeout limits then the number of randomly falling tests would decrease.

When we reverted to an earlier version of Chrome, `127.0.6533.0`, we didn't have any problems.

We are going to continue our investigation tomorrow, but suspect that this issue will be affecting other Chrome users. Perhaps there has been a change that requires to tweaks to our Chrome integration. We will let you know what we find when we have more information to share.

In the meantime, running against an older version and tweaking your timeout defaults should mitigate your issue.

**Comment by massimocode** [comment]
Thanks for looking into this. Good to hear you were able to extract the archive (I had just used 7-zip's GUI on windows to encrypt it) and replicate the flakiness on the latest version of Chrome.

I'm ok for now running against an older version of Chrome. I'll wait to hear how your investigation goes.

**Comment by smcenlly** [comment]
>  Is there a wallaby configuration that I can set which will force wallaby to use separate chrome windows for each worker?

We tried launching a separate Chrome process for each worker today, it fixes the Chrome performance regression and does not introduce any slow-down within Wallaby. We've updated Wallaby to always run this way. This change in behavior is available in Wallaby core `v1.0.1600` and will fix your problem.

By default, Wallaby starts chrome with the `--headless` option. We also discovered that changing this to `--headless=old` also fixed the problem, but given that this option will be removed in the future, we prefer to launch a new process per worker.

---

Thanks again for providing your sample repo; it really helped.
